deps/build.jl

```diff
@@ -0,0 +1,81 @@
+
+using CpuId
+using Libdl
+
+
+function defMKLpath()
+    if Sys.isunix()
+        return "/opt/intel/mkl/lib"
+    # elseif Sys.iswindows()
+    #     return
+    else
+        return ""
+    end
+end
+        
+
+
+### First Option, check for user preference
+if haskey(ENV, "MKL_SL") 
+    basepath = ENV["MKL_SL"]
+    if !any([occursin("libmkl_rt",x) for x in readdir(basepath)])
+        error(""" ENV["MKL_SL"] is set, but does not point to folder containing MKL files """)
+    end
+    println("""Found MKL via ENV["MKL_SL"]
+    basepath set to $basepath""")
+    
+### Next, check if MKL.jl is already installed. 
+# Don't want to add to deps, so bit hacky
+else
+    foundMKLjl = [isdir(joinpath(testdir, "packages/MKL")) for testdir in DEPOT_PATH]
```

Comment:
This does not seem like a good way to detect it. How about:

```
julia> import LinearAlgebra

julia> LinearAlgebra.BLAS.vendor()
:openblas64
```

Reply:
I am not happy with how this works, however I personally have two Julia versions, one with MKL added to the SystemImage, and one without.
With this code I can use VML in either.

I think that might be a special case however, so if you see an issue with this I think that could be handled via the environment variable for me, and be changed to work better for most users.

Reply:
On second thought, and please correct me if I have missed something: After testing with `LinearAlgebra.BLAS.vendor()` I would still have to find the path to the package, as I can not use `Base.find_package` without adding MKL to `[deps]` based on my tests.

Reply:
How about just checking for the existence of `libmkl_rt` with something like

```
julia> using Libdl

julia> dlopen(:libmkl_rt; throw_error=false) !== nothing
false
```

If MKL.jl is installed, or if the `MKL_ROOT` script has been run, I think this should work.

Reply:
Testing on Windows machine with latest MKL.jl installed on Julia 1.4, 
```julia
julia> using LinearAlgebra

julia> BLAS.vendor()
:mkl
```
I get this error that `could not load library "libmkl_rt" The specified module could not be found.`
```julia
Stacktrace:
 [1] dlopen(::String, ::UInt32; throw_error::Bool) at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:109
 [2] dlopen at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:109 [inlined]
 [3] #dlopen#2 at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:105 [inlined]
 [4] dlopen at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:105 [inlined]
 [5] __init__() at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\setup.jl:3
 [6] top-level scope at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\setup.jl:8
 [7] include(::Module, ::String) at .\Base.jl:377
 [8] include(::String) at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\VML.jl:3
 [9] top-level scope at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\VML.jl:12
 [10] include(::Module, ::String) at .\Base.jl:377
 [11] top-level scope at none:2
 [12] eval at .\boot.jl:331 [inlined]
 [13] eval(::Expr) at .\client.jl:457
 [14] top-level scope at .\none:3
in expression starting at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\setup.jl:8
in expression starting at C:\Users\yahyaaba\.julia\packages\VML\f2lRQ\src\VML.jl:12
ERROR: Failed to precompile VML [c8ce9da6-5d36-5c03-b118-5a70151be7bc] to C:\Users\yahyaaba\.julia\compiled\v1.4\VML\WlHoi_Hs8PO.ji.
Stacktrace:
 [1] error(::String) at .\error.jl:33
 [2] compilecache(::Base.PkgId, ::String) at .\loading.jl:1263
 [3] _require(::Base.PkgId) at .\loading.jl:1022
 [4] require(::Base.PkgId) at .\loading.jl:920
 [5] require(::Module, ::Symbol) at .\loading.jl:915
```

Reply:
Also, running what @KristofferC recommends gives the same error:
```julia
julia>dlopen(:libmkl_rt; throw_error=true)
ERROR: could not load library "libmkl_rt"
The specified module could not be found.

Stacktrace:
 [1] dlopen(::String, ::UInt32; throw_error::Bool) at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:109
 [2] #dlopen#2 at C:\Julia-1.4.0-DEV\share\julia\stdlib\v1.4\Libdl\src\Libdl.jl:109 [inlined]
 [3] top-level scope at none:0
```

Reply:
Interesting. I have to admit that I have only tested on unix based systems. I am currently on Windows however, let me have a look.

Reply:
@crown421 I created an issue in MKL https://github.com/JuliaComputing/MKL.jl/issues/18, we had some problems regarding loading MKL on windows because the path did not add to the environment variables, but it was fixed in a PR: 
https://github.com/JuliaComputing/MKL.jl/pull/15/files by @staicfloat
Maybe we should do the same for this dll file too

Reply:
I think I found some solution here in MKL code: https://github.com/JuliaComputing/MKL.jl/blob/1dad6dd032a59d5f9692cf1eb379289b2ed49f0d/src/MKL.jl#L41
```julia
        if Sys.iswindows()
            # On Windows, we have to open the threading library before calling libmkl_rt
            dlopen(libmkl_intel_thread)
        end
```

Reply:
Excellent, once building MKL finishes I will have a look as well.

Reply:
This should be marked as resolved